src/parser/expr.rs

```diff
@@ -27,90 +149,102 @@ impl<'s, 'a> Parser<'s, 'a> {
         Ok(None)
     }
 
-    // disjunction = disjunction '&&' conjunction
-    //             | conjunction
-    fn disjunction(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let mut left = self.conjunction()?;
-        while self.accept_optional(tok![||])?.is_some() {
-            let right = self.conjunction()?;
-            let span = self.spans.merge(&left, &right);
-            left = ExprKind::BinOp(BinOp::LogicalOr, self.alloc(left), self.alloc(right))
-                .with_span(span)
-        }
-        Ok(left)
-    }
-
-    // conjunction = conjunction '&&' inversion
-    //             | inversion
-    fn conjunction(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let mut left = self.comparison()?;
-        while self.accept_optional(tok![&&])?.is_some() {
-            let right = self.comparison()?;
-            let span = self.spans.merge(&left, &right);
-            left = ExprKind::BinOp(BinOp::LogicalAnd, self.alloc(left), self.alloc(right))
-                .with_span(span);
-        }
-        Ok(left)
-    }
+    fn binary_expr(
+        &mut self,
+        mut lhs: Expr<'s, 'a>,
+        last_operator: Option<&Spanned<BinaryOp>>,
+    ) -> ParseResult<(Expr<'s, 'a>, Option<Spanned<BinaryOp>>)> {
+        let precedence_bound = match last_operator {
+            Some(last_operator) => last_operator.precedence(),
+            None => Precedence::Lowest,
+        };
 
-    fn comparison(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let left = self.sum()?;
-        if let Some((op, _span)) = self.accept_any([
-            (tok![==], BinOp::Eq),
-            (tok![!=], BinOp::Neq),
-            (tok![>=], BinOp::Gte),
-            (tok![<=], BinOp::Lte),
-            (tok![>], BinOp::Gt),
-            (tok![<], BinOp::Lt),
-        ])? {
-            let right = self.sum()?;
-            let span = self.spans.merge(&left, &right);
-            return Ok(ExprKind::BinOp(op, self.alloc(left), self.alloc(right)).with_span(span));
-        } else {
-            Ok(left)
+        // maybe we have an operator 'cached' already because we accidentally parsed it
+        // somewhere in the parse loop. At the start, we do not however so None.
+        let mut maybe_operator = None;
+
+        loop {
+            // if we have a cached operator: take it
+            // if we don't, parse a new one!
+            // else, this is the end of the expression
+            let operator = if let Some(operator) = maybe_operator.take() {
+                operator
+            } else if let Some(operator) = self.binary_operator()? {
+                operator
+            } else {
+                break;
+            };
```

Comment:
This can be written much more concisely with `.or_else()`.   In fact, make the whole loop a `while let` based on it.

Reply:
already changed